File: StatusBarSpeedometer/app/src/main/java/ch/rmy/android/statusbar_tacho/utils/Destroyer.java
package ch.rmy.android.statusbar_tacho.utils;

import java.util.ArrayList;
import java.util.List;

public class Destroyer implements Destroyable {

    private final List<Destroyable> destroyables = new ArrayList<>();

    public <T extends Destroyable> T own(T destroyable) {
        destroyables.add(destroyable);
        return destroyable;
    }

    @Override
    public void destroy() {
        for (Destroyable destroyable : destroyables) {
            destroyable.destroy();
        }
        destroyables.clear();
    }

}

File: StatusBarSpeedometer/app/src/main/java/ch/rmy/android/statusbar_tacho/utils/Destroyer.kt
package ch.rmy.android.statusbar_tacho.utils

import java.util.*

class Destroyer : Destroyable {

    private val destroyables = ArrayList<Destroyable>()

    fun <T : Destroyable> own(destroyable: T): T {
        destroyables.add(destroyable)
        return destroyable
    }

    override fun destroy() {
        for (destroyable in destroyables) {
            destroyable.destroy()
        }
        destroyables.clear()
    }

}
